File: Lucy/2025/5-W04/216_Combination_Sum_III.ts
function combinationSum3(k: number, n: number): number[][] {
  const minNum = Array.from({ length: k }, (_, i) => i + 1).reduce((acc, num) => acc + num, 0);
  if (minNum > n) return [];

  const results: number[][] = [];

  const backtrack = (path: number[], currentSum: number, startNumber: number) => {
    if (path.length === k && currentSum === n) {
      results.push([...path]);
      return;
    }

    if (currentSum > n || path.length >= k) {
      return;
    }

    for (let i = startNumber; i <= 9; i++) {
      backtrack([...path, i], currentSum + i, i + 1);
    }
  };

  for (let startNumber = 1; startNumber <= 9; startNumber++) {
    backtrack([startNumber], startNumber, startNumber + 1);
  }

  return results;
}

File: Meli/LeetCode 75/216. Combination Sum III.py
from typing import List

class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        def backtrack(start, path, remaining, results):
            if len(path) == k and remaining == 0:
                results.append(path[:])
                return
            if len(path) >=k or remaining < 0:
                return
            for i in range(start, 10):
                path.append(i)
                backtrack(i+1, path, remaining-i, results)
                path.pop()
            
        results = []
        backtrack(1, [], n, results)
        return results